src/main/java/fi/aalto/cs/apluscourses/ui/base/TextField.java

```diff
@@ -0,0 +1,27 @@
+package fi.aalto.cs.apluscourses.ui.base;
+
+import fi.aalto.cs.apluscourses.ui.utils.TwoWayBindable;
+import java.awt.event.KeyAdapter;
+import java.awt.event.KeyEvent;
+import javax.swing.JTextField;
+
+/**
+ * A class deriving from {@link JTextField} that works with the {@link TwoWayBindable} class to
+ * reflect its text input to the source property.
+ */
+public class TextField extends JTextField {
+  public final transient TwoWayBindable<TextField, String> textBindable =
+      new TwoWayBindable<>(this, TextField::setText, TextField::getText);
```

Comment:
@Taikelenn @jaakkonarhi @stellatsv 

please check this and ensure you understand what is happening here (in terms of the language and also in terms of logcs), ask if unclear

Reply:
I could try to explain this a bit.

So we use Model-View-ViewModel MVVM patttern. That is, we have model classes (that represent our domain objects) and views (pieces of UI) just as in perhaps more familiar MVC pattern. Like in MVC, models should concern only in business logic and be totally ignorant of everything related to presentation of their data. On the other hand, views should be as dumb as possible in the sense that no logic should be located in view classes.

In MVC, we have controllers that work between models and views and take care of the presentation logic. In MVVM, we have view models (a.k.a presentation models) in the place of controllers. View models are, well, models for views: they can be thought as abstract representations of views. For example. if there is a view for personal info and there is text field for person's name then PersonalInfoViewModel has a (string) property "name" whose value is the content of the text field. So, the program doesn't "read" the person's name from the text field but a corresponding property in the view model.

Now, presentation logic (how data is shown, input validation, and so on) can be done in view models that are plain old java classes, and thus, totally unit testable. As said, it's "a representation of the UI" but without any UI framework or widgets messing thing up.

All that a view has to do is to keep its widgets in sync with the view model's properties. This keeping in sync we call _binding_. So we make UI elements' properties (such as text fields' text) bindable in a way presented in this code chunk. Then, in personal info view (in our hypothetical example), we create a binding by calling `nameTextField.bindToSource(viewModel.name)` and that's it (`viewModel.name` is an `ObservableProperty<String>` that is our implementation of observer pattern). Now the view model's "name" property is synced with view's "nameTextField"'s text content so that when one of them changes (either by UI input or by some logic in code), that change is immediately propagated to the other. Neat!

Please feel free to discuss and challenge this approach of ours. And don't hesitate to ask me if there's something that is difficult to understand (I bet there is). If you are not familiar with MVVM pattern and wanna search web there're quite a lot written about it (search either "MVVM" or "Presentation Model". Note that there's a variety in the details of the pattern).